clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -511,17 +511,18 @@ private static int parseAcks(String acksString) {
      * @return The amount of time we waited in ms
      */
     private long waitOnMetadata(String topic, long maxWaitMs) throws InterruptedException {
-        // add topic to metadata topic list if it is not there already.
-        if (!this.metadata.containsTopic(topic))
-            this.metadata.add(topic);
-
+        // add topic to metadata topic list if it is not there already and reset expiry
+        this.metadata.add(topic);
         if (metadata.fetch().partitionsForTopic(topic) != null)
             return 0;
 
         long begin = time.milliseconds();
         long remainingWaitMs = maxWaitMs;
         while (metadata.fetch().partitionsForTopic(topic) == null) {
             log.trace("Requesting metadata update for topic {}.", topic);
+            // Add the topic to metadata before requesting update to ensure that the topic is included
+            // in the metadata topic list since unknown topics are removed from the metadata
+            this.metadata.add(topic);
```

Comment:
@hachikuji I have implemented the code suppressing warnings for UNKNOWN_TOPIC after the first warning. This is fine for the REST proxy scenarios I was trying to fix. I am however concerned that in the general case, it may not be a good idea to suppress warnings for a genuine error. If an application is periodically publishing to a topic that doesn't exist or is manually assigning a partition for consume, at the moment, it gets a bunch of warnings indicating the issue (agree it is too much, but every error corresponds to an attempt to get metadata which adds some load on the client and the broker). Once the warning is suppressed, the load on the client and broker continues because of the tight loop of metadata requests, but the fact that there is a problem is no longer as obvious. The logs are the same regardless of whether there was one UNKNOWN_TOPIC error that resolved itself on the next update or if the problem is a persistent one that never got resolved and the loop of metadata requests is still going on. If we really want to reduce the log spam, I wonder if the warnings should be based on a counter to print 1 in 10 warnings or something like that rather than suppress warnings altogether after the first one. I am beginning to think I should simply use this PR for handling expiry and leave handling of unknown topics for another time.